README.md
# Inkyshot - a shot of inspiration to start the day

![](https://raw.githubusercontent.com/balenalabs-incubator/inkyshot/master/assets/header-photo.jpg)

**Get a daily random inspirational quote delivered direct to your desk with Inkyshot. Build multiple Inkyshots and share the inspiration with your friends, family and loved ones ❤️**

## Hardware required

![](https://raw.githubusercontent.com/balenalabs-incubator/inkyshot/master/assets/hardware-photo.jpg)

* Raspberry Pi (tested with Zero and 3B so far)
* [Pimoroni InkyPHAT display](https://shop.pimoroni.com/products/inky-phat?variant=12549254938707)
* 8GB SD Card (Sandisk Extreme Pro recommended)
* Power supply
* **Optional:** 3D printed case with [micro USB sockets](https://www.aliexpress.com/item/4000484202812.html)
* **Optional:** [Waveshare 2.13" e-paper display V2](https://www.waveshare.com/wiki/2.13inch_e-Paper_HAT) (**Note:** the case designs will need modification to fit this display. Set the `WAVESHARE` environment variable to `1` to enable support)

## Setup & Installation

Running this project is as simple as deploying it to a balenaCloud application. You can do it in just one click by using the button below:

[![](https://balena.io/deploy.png)](https://dashboard.balena-cloud.com/deploy)

You can also deploy in the traditional manner using the balena CLI and `balena push` command. For more information [check out the docs](https://www.balena.io/docs/learn/deploy/deployment/).

## Customization

Your fleet of Inkyshots can all be managed centrally via balenaCloud. Try any of the environment variables below to add some customization.

### Update time

Inkyshot wants to deliver a shot of inspiration to start your day, and by default will do this at 9AM/0900 hours UTC. You can change the hour that the update will happen with the `UPDATE_HOUR` variable; set it anywhere from `0` to `23`.

### Timezone

In order for the update time to work correctly, you'll of course have to tell Inkyshot what timezone you'd like to use. Set the `TZ` environment variable to any [IANA timezone](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), e.g. `Europe/London`, `America/Los_Angeles`, `Asia/Taipei` etc.

### Message override

Is there a special occasion in your family? Override the daily quote with a shot of celebration by setting the `INKY_MESSAGE` environment variable. Try `Happy birthday Sister!`, `Congratulations on the new job, mate!`, or `Happy mother's day!`.

### Font

There are a few fonts built in that you can try. The default is `AmaticSC`, but you can use the `FONT` variable and set it to any of: `FredokaOne`, `HankenGrotesk`, `Intuitive`, `SourceSerifPro`, `SourceSansPro`, `Caladea` and `Roboto`. You're welcome to PR more options into the project!

### Font size

Inkyshot will automatically choose the largest font size for your message that still fits on the display. Too big? Override it with the `FONT_SIZE` variable and Inkyshot will use this as a maximum and continue to resize downwards if your message doesn't fit.

### Test character

When figuring out what size font to use, Inkyshot (invisibly) fills the screen with the `a` character to see how many will fit. You can override this by setting the `TEST_CHARACTER` environment variable with any wider (`w`) or slimmer (`i`) characters of your choosing to adjust this behavior.

### Rotation

If Inkyshot is living in a different house where things aren't necessarily always the same way up, use the `ROTATE` environment variable to rotate the output by 180 degrees.

### Weather
To enable the weather display, set the environment variable `NEXT_DISPLAY` to `weather`.

Next, use either `LATLONG` (e.g. 39.9199,32.8543) or `WEATHER_LOCATION` (e.g. Ankara, Turkey) environment variables to define the location for weather information. Entering only an empty `WEATHER_LOCATION` is also sufficient and in this case Inkyshot will lookup the latitude and longitude information from device's IP address.

Set `SCALE` environment variable to `F` to display the temperature values in Fahrenheit scale. The default is Celcius scale.

Use the `WEATHER_FONT` variable to customize the font used in weather display mode.

`LOCALE` variable allows to display the date of temperature reading in any language supported by [the date library](https://arrow.readthedocs.io/en/latest/#module-arrow.locales).

Set `ALTERNATE_FREQUENCY` to alternate the display mode between quote and weather. Inkyshot uses this environment variable's number value as minutes to update periodically. By default, the first display is quote mode and you can instead chose weather by setting `NEXT_DISPLAY` to `weather`.

Finally, you can choose to fix Inkyshot to display only weather. Set `NEXT_DISPLAY` to `weather`, `DISPLAY_QUOTE` to false and don't set `INKY_MESSAGE`. In this setup, inkyshot will keep updating weather information after every minutes set by `ALTERNATE_FREQUENCY`.


## Case

STL files are included within the assets folder of the project for you to 3D print your own case.

![](https://raw.githubusercontent.com/balenalabs-incubator/inkyshot/master/assets/inky-print.png)

The case has two positions for a captive M3 nut, and can be fastened together with two countersunk 8mm M3 machine screws.

A position is open in the rear of the case for the use of a [micro USB PCB socket](https://www.aliexpress.com/item/4000484202812.html), allowing for direct connection of power to the back of a Raspberry Pi Zero.

![](https://raw.githubusercontent.com/balenalabs-incubator/inkyshot/master/assets/inky-rear.png)


## Credits

Quotes are delivered from the
<span style="z-index:50;font-size:0.9em; font-weight: bold;">
      <img src="https://theysaidso.com/branding/theysaidso.png" height="20" width="20" alt="theysaidso.com"/>
      <a href="https://theysaidso.com" title="Powered by quotes from theysaidso.com" style="color: #ccc; margin-left: 4px; vertical-align: middle;">
        They Said So®
      </a>
</span> REST API.

Location information based on IP address is retrieved from [ipinfo.io](https://ipinfo.io).

We use Norwegian Meteorological Institute's [weather API](https://api.met.no/weatherapi/locationforecast/2.0/documentation) to get the weather forecast of the day. Weather symbols are customized from [yr.no's repository](https://github.com/nrkno/yr-weather-symbols).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v1.0.0
## (2020-09-15)

* Display weather forecast with icons [gelbal]

# v0.2.3
## (2020-09-09)

* If message is blank, use device name [Chris Crocker-White]

# v0.2.2
## (2020-06-19)

* Add extra logging of message and font size [Chris Crocker-White]

# v0.2.1
## (2020-06-16)

* Update README with Waveshare support [Chris Crocker-White]

# v0.2.0
## (2020-06-16)

* Add libraries for Waveshare 2.13" display [Chris Crocker-White]
* Added support for Waveshare e-paper. [Larry Matter]

# v0.1.0
## (2020-06-07)

* Add absolute font path [Chris Crocker-White]
* Organize and add more case designs [Chris Crocker-White]
* Add pixel font for long messages [Chris Crocker-White]
* Add hardware image to readme [Chris Crocker-White]
* Add a timeout and handle it [Chris Crocker-White]
* Update readme to mention rotation [Chris Crocker-White]
* Housekeeping, add readme, ignore files [Chris Crocker-White]
* Handle HTTP errors on quote load [Chris Crocker-White]
* Add scheduler [Chris Crocker-White]
* Assume a default font (now all vars optional) [Chris Crocker-White]
* Allow font size override [Chris Crocker-White]
* Calculate font size automatically [Chris Crocker-White]
* Simplify and change to multiline_text [Chris Crocker-White]
* Rename, reorganize and comment [Chris Crocker-White]
* Added rotation and default variables [Catherine Tung]
* Set to pull quote of the day from API [Chris Crocker-White]
* Use wheels for Pi Zero (Pi 3 for dev) [Chris Crocker-White]
* i Added fonts and font as a environment variable [Catherine Tung]
* Text Wrapping and font sizing [Catherine Tung]
* New font addition Hanken Grotesk [Catherine Tung]
* Intial commit Set up Dockerfile with Python packages [Catherine Tung]


start.sh
#!/bin/bash
# Get the current device name
export DEVICE_NAME=$(curl -sX GET "https://api.balena-cloud.com/v5/device?\$filter=uuid%20eq%20'$BALENA_DEVICE_UUID'" \
-H "Content-Type: application/json" \
-H "Authorization: Bearer $BALENA_API_KEY" | \
jq -r ".d | .[0] | .device_name")

# Run the display update once on container start
python /usr/app/update-display.py

# Save out the current env to a file so cron job can use it
export -p > /usr/app/env.sh

# Set default values if these env vars are not set
if [[ -z "${ALTERNATE_FREQUENCY}" ]]; then
  Alternate="0"
  [[ -z "${UPDATE_HOUR}" ]] && UpdateHour='9' || UpdateHour="${UPDATE_HOUR}"
else
  Alternate="*/${ALTERNATE_FREQUENCY}"
  UpdateHour='*'
fi

# Add the job to the crontab using update_hour var, defaulting to 9 AM
(echo "${Alternate} ${UpdateHour} * * * /usr/app/run-update.sh") | crontab -

# Start the cron daemon as PID 1
exec cron -f

update-display.py
import json
import logging
import math
import os
from os import environ
import sys
import textwrap
import time
import urllib.request
import urllib.error

from font_amatic_sc import AmaticSC
from font_caladea import Caladea
from font_fredoka_one import FredokaOne
from font_hanken_grotesk import HankenGrotesk
from font_intuitive import Intuitive
from font_roboto import Roboto
from font_source_sans_pro import SourceSansPro
from font_source_serif_pro import SourceSerifPro
from PIL import Image, ImageFont, ImageDraw
import arrow
import geocoder

icon_map = {
    "clearsky": 1,
    "cloudy": 4,
    "fair": 2,
    "fog": 15,
    "heavyrain": 10,
    "heavyrainandthunder": 11,
    "heavyrainshowers": 41,
    "heavyrainshowersandthunder": 25,
    "heavysleet": 48,
    "heavysleetandthunder": 32,
    "heavysleetshowers": 43,
    "heavysleetshowersandthunder": 27,
    "heavysnow": 50,
    "heavysnowandthunder": 34,
    "heavysnowshowers": 45,
    "heavysnowshowersandthunder": 29,
    "lightrain": 46,
    "lightrainandthunder": 30,
    "lightrainshowers": 40,
    "lightrainshowersandthunder": 24,
    "lightsleet": 47,
    "lightsleetandthunder": 31,
    "lightsleetshowers": 42,
    "lightsnow": 49,
    "lightsnowandthunder": 33,
    "lightsnowshowers": 44,
    "lightssleetshowersandthunder": 26,
    "lightssnowshowersandthunder": 28,
    "partlycloudy": 3,
    "rain": 9,
    "rainandthunder": 22,
    "rainshowers": 5,
    "rainshowersandthunder": 6,
    "sleet": 12,
    "sleetandthunder": 23,
    "sleetshowers": 7,
    "sleetshowersandthunder": 20,
    "snow": 13,
    "snowandthunder": 14,
    "snowshowers": 8,
    "snowshowersandthunder": 21,
}

def create_mask(source):
    """Create a transparency mask to draw images in grayscale
    """
    logging.info("Creating a transparency mask for the image")
    mask_image = Image.new("1", source.size)
    w, h = source.size
    for x in range(w):
        for y in range(h):
            p = source.getpixel((x, y))
            if p in [BLACK, WHITE]:
                mask_image.putpixel((x, y), 255)
    return mask_image

def draw_weather(weather, img, scale):
    """Draw the weather info on screen"""
    logging.info("Prepare the weather data for drawing")
    # Draw today's date on left side below today's name
    today = arrow.utcnow().format(fmt="DD MMMM", locale=LOCALE)
    date_font = ImageFont.truetype(WEATHER_FONT, 18)
    draw.text((3, 3), today, BLACK, font=date_font)
    # Draw current temperature to right of today
    temp_font = ImageFont.truetype(WEATHER_FONT, 24)
    draw.text((3, 30), f"{temp_to_str(weather['temperature'], scale)}°", BLACK, font=temp_font)
    # Draw today's high and low temps on left side below date
    small_font = ImageFont.truetype(WEATHER_FONT, 14)
    draw.text(
        (3, 72),
        f"{temp_to_str(weather['min_temp'], scale)}° - {temp_to_str(weather['max_temp'], scale)}°",
        BLACK,
        font=small_font,
    )
    # Draw today's max humidity on left side below temperatures
    draw.text((3, 87), f"{weather['max_humidity']}%", BLACK, font=small_font)
    # Load weather icon for day or night
    icon_name = weather['symbol'].split('_')[0]
    time_of_day = 'd' if weather['symbol'].split('_')[1] == 'day' else 'n'
    icon_filename = f"{icon_map[icon_name]:02}{time_of_day}.png"
    icon_image = Image.open(f"./weather-icons/{icon_filename}")
    #icon_mask = create_mask(icon_image)
    # Draw the weather icon
    #img.paste(icon_image, (120, 3), icon_mask)
    img.paste(icon_image, (120, 3))
    return img

def get_location():
    """Return coordinate and location info based on IP address"""
    url = "https://ipinfo.io"
    req = urllib.request.Request(url, headers={"Accept" : "application/json"})
    try:
        res = urllib.request.urlopen(req, timeout=5)
        logging.info("Retrieved the location data using device's IP address")
        if(res.status == 200):
            return json.loads(res.read().decode())
    except (urllib.error.HTTPError, urllib.error.URLError) as err:
        logging.error(err)
    logging.error("Failed to retrieve the location data")
    return {}

def get_weather(lat, long):
    """Return weather report for the next 24 hours"""
    url = f"https://api.met.no/weatherapi/locationforecast/2.0/compact?lat={lat}&lon={long}"
    req = urllib.request.Request(url, headers={"Accept" : "application/json"})
    logging.info("Retrieving weather forecast")
    weather = {}
    try:
        res = urllib.request.urlopen(req, timeout=5)
        if(res.status == 200):
            data = json.loads(res.read().decode())
            timeseries = data['properties']['timeseries']
            now = arrow.utcnow()
            tomorrow = now.shift(hours=+24)
            weather_24hours = []
            for t in timeseries:
                tm = arrow.get(t['time'])
                if tm < tomorrow:
                    temp = t['data']['instant']['details']['air_temperature']
                    humid = t['data']['instant']['details']['relative_humidity']
                    symbol = t['data']['next_1_hours']['summary']['symbol_code']
                    weather_24hours.append({
                        'time': tm,
                        'temperature': temp,
                        'humidity': humid,
                        'symbol': symbol,
                    })
            weather_24hours = sorted(weather_24hours, key=lambda x: x['time'])
            weather = [x for x in weather_24hours if x['time'] <= now.shift(hours=+1)][-1]
            temperatures = [x['temperature'] for x in weather_24hours if x['time'] <= now.shift(days=+1)]
            weather['max_temp'] = max(temperatures)
            weather['min_temp'] = min(temperatures)
            weather['max_humidity'] = max([x['humidity'] for x in weather_24hours if x['time'] <= now.shift(days=+1)])
    except (urllib.error.HTTPError, urllib.error.URLError) as err:
        logging.error(err)
    return weather

def temp_to_str(temp, scale):
    """Prepare the temperature to draw based on the defined scale: Celcius or Fahrenheit"""
    if scale == 'F':
        temp = temp * 9/5 + 32
    return f"{temp:.1f}"

# Read the preset environment variables and overwrite the default ones
if "DEBUG" in os.environ:
    logging.basicConfig(level=logging.DEBUG)
else:
    logging.basicConfig(level=logging.INFO)

# Check if daily quote is disabled
DISPLAY_QUOTE = True
if "DISPLAY_QUOTE" in os.environ and not os.environ["DISPLAY_QUOTE"]:
    DISPLAY_QUOTE = False

# Assume a default font if none set
FONT_SELECTED = AmaticSC
if "FONT" in os.environ:
    FONT_SELECTED = locals()[os.environ["FONT"]]

FONT_SIZE = 24
if "FONT_SIZE" in os.environ:
    FONT_SIZE = int(os.environ["FONT_SIZE"])

FONT = ImageFont.truetype(FONT_SELECTED, FONT_SIZE)

WEATHER_FONT = FredokaOne
if "WEATHER_FONT" in os.environ:
    WEATHER_FONT = locals()[os.environ["WEATHER_FONT"]]

[LAT, LONG] = [float(x) for x in os.environ["LATLONG"].split(",")] if "LATLONG" in os.environ else [None, None]

# Temperature scale
SCALE = 'F' if "SCALE" in os.environ and os.environ["SCALE"] == 'F' else 'C'

# Locale formatting of date
LOCALE = os.environ["LOCALE"] if "LOCALE" in os.environ else 'en'

# Alternate displaying quote and weather
NEXT_DISPLAY = os.environ["NEXT_DISPLAY"] if "NEXT_DISPLAY" in os.environ else 'quote'

# Init the display. TODO: support other colours
logging.debug("Init and Clear")
if "WAVESHARE" in os.environ:
    logging.info("Display type: Waveshare")

    import lib.epd2in13_V2
    display = lib.epd2in13_V2.EPD()
    display.init(display.FULL_UPDATE)
    display.Clear(0xFF)
    # These are the opposite of what InkyPhat uses.
    WIDTH = display.height # yes, Height
    HEIGHT = display.width # yes, width
    BLACK = "black"
    WHITE = "white"
    img = Image.new('1', (WIDTH, HEIGHT), 255)
else:
    logging.info("Display type: InkyPHAT")

    from inky import InkyPHAT
    display = InkyPHAT("black")
    display.set_border(display.WHITE)
    WIDTH = display.WIDTH
    HEIGHT = display.HEIGHT
    BLACK = display.BLACK
    WHITE = display.WHITE
    img = Image.new("P", (WIDTH, HEIGHT))

draw = ImageDraw.Draw(img)

logging.info("Display dimensions: W %s x H %s", WIDTH, HEIGHT)

display_weather = False

if NEXT_DISPLAY == 'weather':
    weather_location = None
    if "WEATHER_LOCATION" in os.environ:
        weather_location = os.environ["WEATHER_LOCATION"]
    # Get the latitute and longitude of the address typed in the env variable if latitude and longitude are not set
    if weather_location and (not LAT or not LONG):
        logging.info(f"Location is set to {weather_location}")
        try:
            geo = geocoder.arcgis(weather_location)
            [LAT, LONG] = geo.latlng
        except Exception as e:
            print(f"Unexpected error: {e.message}")

    # If no address or latitute / longitude are found, retrieve location via IP address lookup
    if not LAT or not LONG:
        location = get_location()
        [LAT, LONG] = [float(x) for x in location['loc'].split(',')]
    weather = get_weather(LAT, LONG)
    # Set latitude and longituted as environment variables for consecutive calls
    os.environ['LATLONG'] = f"{LAT},{LONG}"
    # If weather is empty dictionary fall back to drawing quote
    if len(weather) > 0:
        img = draw_weather(weather, img, SCALE)
        display_weather = True

message = None
# Use a dashboard defined message if we have one, otherwise load a nice quote
if "INKY_MESSAGE" in os.environ and not display_weather and DISPLAY_QUOTE:
   message = os.environ['INKY_MESSAGE']

   if message == "":
       # If the message var was set but blank, use the device name
       message = os.environ['DEVICE_NAME']
elif not display_weather and DISPLAY_QUOTE:
    req = urllib.request.Request("https://quotes.rest/qod?language=en", headers={"Accept" : "application/json"})
    try:
        res = urllib.request.urlopen(req, timeout=5).read()
        data = json.loads(res.decode())
        message = data['contents']['quotes'][0]['quote']
    except (urllib.error.HTTPError, urllib.error.URLError) as err:
        FONT_SIZE = 25
        message = "Sorry folks, today's quote has gone walkies :("

if message:
    logging.info("Message: %s", message)
    # Work out what size font is required to fit this message on the display
    message_does_not_fit = True

    test_character = "a"
    if "TEST_CHARACTER" in os.environ:
        test_character = os.environ['TEST_CHARACTER']

    while message_does_not_fit == True:
        test_message = ""
        message_width = 0
        FONT_SIZE -= 1

        if FONT_SIZE <= 17:
            FONT_SIZE = 8
            FONT = ImageFont.truetype("/usr/app/fonts/Grand9KPixel.ttf", FONT_SIZE)

        # We're using the test character here to work out how many characters
        # can fit on the display when using the chosen font
        while message_width < WIDTH:
            test_message += test_character
            message_width, message_height = draw.textsize(test_message, font=FONT)

        max_width = len(test_message)
        max_lines = math.floor(HEIGHT/message_height)

        # We wrap the message to the width we worked out earlier
        wrapper = textwrap.TextWrapper(width=max_width)
        word_list = wrapper.wrap(text=message)

        if len(word_list) <= max_lines:
            message_does_not_fit = False

        if FONT_SIZE < 9:
            message_does_not_fit = False

    logging.info("Font size: %s", FONT_SIZE)
    offset_x, offset_y = FONT.getoffset(message)

    # Rejoin the wrapped lines with newline chars
    separator = '\n'
    output_text = separator.join(word_list)

    w, h = draw.multiline_textsize(output_text, font=FONT, spacing=0)

    x = (WIDTH - w)/2
    y = (HEIGHT - h - offset_y)/2
    draw.multiline_text((x, y), output_text, BLACK, FONT, align="center", spacing=0)

# Rotate and display the image
if "ROTATE" in os.environ:
    img = img.rotate(180)

if "WAVESHARE" in os.environ:
    # epd does not have a set_image method.
    display.display(display.getbuffer(img))
else:
    display.set_image(img)
    display.show()

logging.info("Done drawing")
# Alternate to other display
os.environ["NEXT_DISPLAY"] = 'weather' \
    if (NEXT_DISPLAY == 'quote' and "ALTERNATE_FREQUENCY" in os.environ) or not DISPLAY_QUOTE \
    else 'quote'

sys.exit(0)
